File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.js
function solution(players, callings) {
    let rank = {};
    let rankName = {};
    let answer = [];
    for (let i=0; i<players.length; i++) {
        rank[i] = players[i];
        rankName[players[i]] = i;
    }
    for (name of callings) {
        let curRank = rankName[name];
        let curFaster = rank[curRank-1];
        rank[curRank] = curFaster;
        rank[curRank-1] = name;
        rankName[curFaster] = curRank;
        rankName[name] = curRank-1;
    }
    answer = Object.values(rank);
    return answer;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        String[] answer = {};
        HashMap<String, Integer> ranking = new HashMap<String, Integer>();
        for (int i=0; i<players.length; i++) {
            ranking.put(players[i], i);
        }
        for (String name: callings) {
            int curRank = ranking.get(name);
            String temp = players[curRank-1];
            ranking.put(name, ranking.get(name) - 1);
            ranking.put(temp, ranking.get(temp) + 1);
            players[curRank-1] = players[curRank];
            players[curRank] = temp;
        }
        answer = players.clone();
        return answer;
    }
}